### ISSUE TITLE
The server unexpectedly closing a connection bricks a connection pool

### ISSUE THREAD

**Issue by lucacasonato** [denodrivers/postgres#355]
Here is a test case that reproduces the issue:

```ts
Deno.test("Attempts reconnection when DB goes away unexpectedly", async function () {
  const cfg = getMainConfiguration();

  const aborter = new AbortController();
  const listener = Deno.listen({ port: 0 });

  const proxy = (async () => {
    for await (const conn of listener) {
      const outbound = await Deno.connect({
        hostname: cfg.hostname,
        port: Number(cfg.port),
      });
      aborter.signal.addEventListener("abort", () => {
        conn.close();
        outbound.close();
      });
      await Promise.all([
        copy(conn, outbound),
        copy(outbound, conn),
      ]).catch(() => {});
    }
  })();

  const pool = new Pool({
    ...cfg,
    port: (listener.addr as Deno.NetAddr).port,
  }, 1);

  const client = await pool.connect();
  await client.queryObject("SELECT 1");
  client.release();

  // This closes ongoing connections. The original connection is now dead, so
  // a new connection should be established.
  aborter.abort();

  const client2 = await pool.connect();
  await client2.queryObject("SELECT 1");
  client2.release();

  listener.close();
  await proxy;
});
```

While this is an artificial reproduction, I have hit this in the wild for connections to supabase.io servers. They seem to just disconnect after a while, causing the `Pool` in a state where whatever I do, I get one of these two errors (note the stack traces are not correct, because of a Deno Deploy bug):

```
ConnectionReset: Connection reset by peer (os error 104)
    at async read (deno:ext/net/01_net.js:21:19)
    at async BufReader.read (https://deno.land/std@0.114.0/io/buffer.ts:314:18)
    at async BufReader.readFull (https://deno.land/std@0.114.0/io/buffer.ts:343:28)
    at async Connection.#readMessage (https://deno.land/x/postgres@v0.14.2/connection/connection.ts:95:9)
    at async Connection.#simpleQuery (https://deno.land/x/postgres@v0.14.2/connection/connection.ts:451:31)
    at async Connection.query (https://deno.land/x/postgres@v0.14.2/connection/connection.ts:650:24)
    at async Server.<anonymous> (file:///src/main.tsx:13:28)
    at async Server.#respond (https://deno.land/std@0.114.0/http/server.ts:246:30)

BrokenPipe: Broken pipe (os error 32)
    at async write (deno:ext/net/01_net.js:26:12)
    at async BufWriter.flush (https://deno.land/std@0.114.0/io/buffer.ts:633:29)
    at async Connection.#simpleQuery (https://deno.land/x/postgres@v0.14.2/connection/connection.ts:443:9)
    at async Connection.query (https://deno.land/x/postgres@v0.14.2/connection/connection.ts:650:24)
    at async Server.<anonymous> (file:///src/main.tsx:13:28)
    at async Server.#respond (https://deno.land/std@0.114.0/http/server.ts:246:30)
```

**Comment by iugo** [comment]
@Soremwar same error with postgres@v0.15.0:

```
ConnectionReset: Connection reset by peer (os error 54)
    at async read (deno:ext/net/01_net.js:21:19)
    at async BufReader.read (https://deno.land/std@0.114.0/io/buffer.ts:383:12)
    at async BufReader.readFull (https://deno.land/std@0.114.0/io/buffer.ts:415:20)
    at async Connection.#readMessage (https://deno.land/x/postgres@v0.15.0/connection/connection.ts:152:5)
    at async Connection.#preparedQuery (https://deno.land/x/postgres@v0.15.0/connection/connection.ts:853:27)
    at async Connection.query (https://deno.land/x/postgres@v0.15.0/connection/connection.ts:930:16)
    at async routerHandler (file:///Users/zheren/git/z1p/src/pages/product/spu-cate-list/get.ts:5:15)
    at async Server.handler (file:///Users/zheren/git/z1p/src/main.ts:14:37)
    at async Server.#respond (https://deno.land/std@0.122.0/http/server.ts:298:18)
BrokenPipe: Broken pipe (os error 32)
    at async write (deno:ext/net/01_net.js:26:12)
    at async BufWriter.flush (https://deno.land/std@0.114.0/io/buffer.ts:745:21)
    at async Connection.#preparedQuery (https://deno.land/x/postgres@v0.15.0/connection/connection.ts:843:5)
    at async Connection.query (https://deno.land/x/postgres@v0.15.0/connection/connection.ts:930:16)
```

**Comment by Soremwar** [comment]
@iugo Are you experiencing this problem with the reproduction that lucacasonato provided?

**Comment by iugo** [comment]
> @iugo Are you experiencing this problem with the reproduction that lucacasonato provided?

I encountered this problem in a new project I was using.

Whenever I was idle for a period of time, I would get an error when connecting again.

My current solution is, `await pool.end()` after the error occurs.

**Comment by dylanpyle** [comment]
FYI, still seeing this on Broken Pipe errors, same output as https://github.com/denodrivers/postgres/issues/436

**Comment by dylanpyle** [comment]
It looks like the fix in #365 handles ConnectionErrors in the `query` case, but not in others - e.g. `transaction.begin()` doesn't seem to go through that path, so if it hits an error it stays dead. Here's an example stack trace of an error my app did not recover from:

```
ConnectionError: The session was terminated unexpectedly
at Connection.#readMessage (https://deno.land/x/postgres@v0.19.2/connection/connection.ts:186:13)
at eventLoopTick (ext:core/01_core.js:169:7)
at async Connection.#simpleQuery (https://deno.land/x/postgres@v0.19.2/connection/connection.ts:690:27)
at async Connection.query (https://deno.land/x/postgres@v0.19.2/connection/connection.ts:977:18)
at async PoolClient.#executeQuery (https://deno.land/x/postgres@v0.19.2/client.ts:256:12)
at async PoolClient.queryArray (https://deno.land/x/postgres@v0.19.2/client.ts:339:12)
at async Transaction.begin (https://deno.land/x/postgres@v0.19.2/query/transaction.ts:254:7)

```

**Comment by asciiwhite** [comment]
I also have an similar issue with a BrokenPipe with v0.19.5 and Deno 2.3.6.

`BrokenPipe: Broken pipe (os error 32)
    at Object.write (ext:deno_web/06_streams.js:1181:20)
    at Module.invokeCallbackFunction (ext:deno_webidl/00_webidl.js:1116:16)
    at WritableStreamDefaultController.writeAlgorithm (ext:deno_web/06_streams.js:3952:14)
    at writableStreamDefaultControllerProcessWrite (ext:deno_web/06_streams.js:4522:55)
    at writableStreamDefaultControllerAdvanceQueueIfNeeded (ext:deno_web/06_streams.js:4425:5)
    at writableStreamDefaultControllerWrite (ext:deno_web/06_streams.js:4569:3)
    at writableStreamDefaultWriterWrite (ext:deno_web/06_streams.js:4716:3)
    at WritableStreamDefaultWriter.write (ext:deno_web/06_streams.js:6618:12)
    at Connection.#appendArgumentsToMessage (file:///home/cvix/workspace/postgres/connection/connection.ts:829:30)
    at Connection.#preparedQuery (file:///home/cvix/workspace/postgres/connection/connection.ts:888:41)
    at Connection.#query(file:///home/cvix/workspace/postgres/connection/connection.ts:980:41)
`
The catch block in Connection.query() does not handle the BrokenPipe error and so does not end the connectionm which is important for an correct startup() call for the following query() calls. So a reconnection is not handled correctly.

A local fix inside Connection.query() and Connection.end() solved the issue.

Connection.query()

`catch (e) {
      if (e instanceof ConnectionError ||
          e instanceof Deno.errors.BrokenPipe) {
        await this.end();
      }
      throw e;
}`

Connection.end()

`if (this.connected) {
      const terminationMessage = new Uint8Array([0x58, 0x00, 0x00, 0x00, 0x04]);
      try {
        await this.#connWritable.write(terminationMessage);
        await this.#connWritable.ready;
      } catch (_e) {
        // This steps can fail if the underlying connection was closed ungracefully
      } finally {
        this.#closeConnection();
        this.#onDisconnection();
      }
 }`
